File: vending_machine/vending.cpp
#include <iostream>
#include <string>
#include <math.h>

using namespace::std;

void fill_price_array(int num, float price_array[]){
	for(int i = 0; i < num; i++){
		float price = (rand() % 10) * 0.25;
		if(price == 0){
			price = 0.5;
		}

		price_array[i] = roundf(price * 100) / 100;
	}
}





void fill_item_array(int num, string item_array[]){
	int l_count;
	char letters[5] = {'A', 'B', 'C', 'D', 'E'};
	string item;
	for(int i=0; i<num; i++){
		l_count = floor(i/5);
		item = letters[l_count] + to_string( (i%5) + 1 );
		item_array[i] = item;
	}
}


string make_whitespace(int len_diff){
	string whitespace = "";
	for(int i=0; i<len_diff; i++){
		whitespace += " ";
	}
	return whitespace;
}


void print_vending_machine(int num, float price_array[], string item_array[]){
	int len_diff;

	for(int i=1; i<num; i++){
		if( ((i+1)%5) == 0){
			string price_line[5];
			string item_line[5];
			for(int j = i-4; j<i+1; j++){
				item_line[j%5] = item_array[j];
				price_line[j%5] = "$" + to_string(price_array[j]);
				if(item_array[j] != "SOLD OUT"){
					len_diff = price_line[j%5].length() - item_line[j%5].length();
					item_line[j%5] += make_whitespace(len_diff);
				}
				else{
					len_diff = 8 - price_line[j%5].length();
					price_line[j%5] += make_whitespace(len_diff);
				}
			}

		for(int i_count = 0; i_count<5; i_count++){
			cout << item_line[i_count] + " || ";
		}
		cout << endl;
		for(int p_count=0; p_count<5; p_count++){
			cout << price_line[p_count] + " || ";
		}
		cout << endl;
		cout << endl;
		}
	}
}


bool is_sold_out(int num, bool sold_out_list[]){
	for(int i=0; i<num; i++){
		if(sold_out_list[i] != true){
			return false;
		}
	}

	return true;
}



int get_index_of_items(int num, string choice, string item_array[]){
	for(int i = 0; i<num; i++){
		if(item_array[i] == choice){
			return i;
		}
	}
	return 404; // code that something's wrong
}



int main(){

	int num;

	cout << "How large would you like the machine to be?";
	cout << " ";
	cout << "Please keep the size of the machine <= 25." << endl;
	cin >> num;

	// num error
	if(num > 25){
		cout << "You can't have more than 25 items in the machine." << endl;
		throw "num_error";
	}


	float price_array[num];
	string item_array[num];



	fill_price_array(num, price_array);
	fill_item_array(num, item_array);

	float money;
	cout << "How much money do you have?" << endl;
	cin >> money;

	bool sold_out_list[num];
	string remaining_money;
	string choice;
	// string item_array_titles = item_array;

	while(money > 0){

		if(is_sold_out(num, sold_out_list) == true){
			cout << "This machine is empty!" << endl;
			remaining_money = "$" + to_string(money);
			cout << "You have " + remaining_money + " money" << endl;
			return 0;
		}

		remaining_money = "$" + to_string(money);
		cout << "You have " + remaining_money + " remaining." << endl;

		print_vending_machine(num, price_array, item_array);

		cout << endl << "What's your choice?" << endl;
		cin >> choice;

		if(get_index_of_items(num, choice, item_array) == 404){
			cout << "This item is sold out! Please choose another item." << endl;
			continue;
		}

		// if(item_array[get_index_of_items(num, choice, item_array)] == "SOLD_OUT"){
			// cout << "Please choose a valid selection." << endl;
			// continue;
		// }

		if(money < price_array[get_index_of_items(num, choice, item_array)]){
			cout << "You can't afford that! Please choose another item." << endl;
			continue;
		}

		money -= price_array[get_index_of_items(num, choice, item_array)];

		sold_out_list[get_index_of_items(num, choice, item_array)] = true;

		item_array[get_index_of_items(num, choice, item_array)] = "SOLD OUT";



		// TODO: make function to find index of a given choice, change it to
		// 'SOLD OUT' and add 'true'

	}




	return 0;
}

File: vending_machine/vending.java
import java.util.Scanner;
import java.util.ArrayList;
import java.lang.Math;

class vending {

	static ArrayList<Double> make_prices(int num, int lower_price_limit, int upper_price_limit) {

		ArrayList<Double> prices = new ArrayList<Double>();
		for(int i=0; i<num; i++){
			prices.add(Math.floor(Math.random() * 10) * 0.25);
		}

		return prices;
	}



	static ArrayList<String> make_items(ArrayList<Double> prices_list) {

		ArrayList<String> items_list = new ArrayList<String>();
		char[] letters = new char[]{'A', 'B', 'C', 'D', 'E'};

		for(int n = 0; n < prices_list.size(); n++) {
			int l_count = (int)Math.floor(n / 5);
			String item = letters[l_count] + String.valueOf(n%5 + 1);
			items_list.add(item);
		}

		return items_list;

	}


	static String make_whitespace(int len_diff) {

		String whitespace = new String("");
		for(int i=0; i<len_diff; i++){
			whitespace += " ";
		}
		
		return whitespace;
	}



	static void print_vending_machine(int num, ArrayList<Double> prices_list, ArrayList<String> items_list) {
	
		int len_diff;

		for(int i=1; i<num; i++){

			if(( (i+1)%5 ) == 0){

				String[] price_line = new String[5];
				String[] item_line = new String[5];
				for(int j = i-4; j<i+1; j++){
					item_line[j%5] = items_list.get(j);
					price_line[j%5] = "$" + String.valueOf(prices_list.get(j));
					if(items_list.get(j) != "SOLD OUT"){
						len_diff = price_line[j%5].length() - item_line[j%5].length();
						item_line[j%5] += make_whitespace(len_diff);
					}
					else {
						len_diff = 8 - price_line[j%5].length();
						price_line[j%5] += make_whitespace(len_diff);
					}
				}

			for(int i_count = 0; i_count<5; i_count++){
				System.out.print(item_line[i_count] + " || ");
			}

			System.out.println();

			for(int p_count=0; p_count<5; p_count++){
				System.out.print(price_line[p_count] + " || ");
			}

			System.out.println("");
			System.out.println("");

			}
		}
	}



	public static void main(String[] args) {
	
		System.out.println("How many entries would you like in your vending machine?" + "\n" +
	    "Please keep the number of entries <= 25.");

		Scanner scan = new Scanner(System.in);
		Scanner scan_str = new Scanner(System.in);
		int num = scan.nextInt();

		if(num > 25){
			System.out.println("You can't have more than 25 items in the machine.");
			throw new RuntimeException("num_error");
		}

		ArrayList<Double> prices_list = make_prices(num, 1, 8);
		ArrayList<String> items_list = make_items(prices_list);

		System.out.println("How much money do you have?");
		int money_in = scan.nextInt();

		ArrayList<String> sold_out_list = new ArrayList<String>();
		String remaining_money = new String();
		String choice = new String();

		// cast money var to double
		float money = (float)money_in;


		while(money > 0) {

			if(sold_out_list.size() == items_list.size()){
				System.out.println("This machine is empty!");
				remaining_money = "$" + String.valueOf(money);
				System.out.println("You have " + remaining_money + " money.");
				return;
			}

			remaining_money = "$" + String.valueOf(money);
			System.out.println("You have " + remaining_money + " money.");

			print_vending_machine(num, prices_list, items_list);

			System.out.println("What's your choice?");
			choice = scan_str.nextLine();

			if(sold_out_list.contains(choice)){
				System.out.println("This item is sold out! Please choose another item.");
				continue;
			}

			if(money < prices_list.get( items_list.indexOf(choice) )){
				System.out.println("You can't afford that! Please choose another item.");
				continue;
			}

			money -= prices_list.get(items_list.indexOf(choice));

			sold_out_list.add(choice);

			items_list.set(items_list.indexOf(choice), "SOLD OUT");


			}



	}
}